docs/reference/modules/scripting/painless-debugging.asciidoc

```diff
@@ -0,0 +1,117 @@
+
+Painless doesn't have a
+https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop[REPL]
+and while it'd be nice for it to have one one day, it wouldn't tell you the
+whole story around debugging painless scripts embedded in Elasticsearch because
+the data that the scripts have access to or "context" is so important. For now
+the best way to debug embedded scripts is by throwing exceptions at choice
+places. While you can throw your own exceptions
+(`throw new Exception('whatever')`), Painless's sandbox prevents you from
+accessing useful information like the type of an object. So Painless has a
+utility method, `Debug.explain` which throws the exception for you. For
+example, you can use the <<search-explain>> to explore the context available to
+a <<query-dsl-script-query>>.
+
+[source,js]
+---------------------------------------------------------
+PUT /hockey/player/1?refresh
+{"first":"johnny","last":"gaudreau","goals":[9,27,1],"assists":[17,46,0],"gp":[26,82,1]}
+
+POST /hockey/player/1/_explain
+{
+  "query": {
+    "script": {
+      "script": "Debug.explain(doc.goals)"
+    }
+  }
+}
+---------------------------------------------------------
+// CONSOLE
+// TEST[catch:/painless_explain_error/]
+
+Which shows that the class of `doc.first` is
+`org.elasticsearch.index.fielddata.ScriptDocValues.Longs` by responding with:
+
+[source,js]
+---------------------------------------------------------
+{
+   "error": {
+      "root_cause": ...,
+      "type": "script_exception",
+      "reason": "runtime error",
+      "caused_by": {
+         "type": "painless_explain_error",
+         "reason": "getClass=[org.elasticsearch.index.fielddata.ScriptDocValues$Longs] toString=[[1, 9, 27]]" <1>
+      },
+      ...
+   },
+   "status": 500
+}
+---------------------------------------------------------
+// TESTRESPONSE[s/"root_cause": \.\.\./"root_cause": $body.error.root_cause/]
+// TESTRESPONSE[s/\.\.\./"script_stack": $body.error.script_stack, "script": $body.error.script, "lang": $body.error.lang/]
+
+<1> This is the explanation.
+
+You can use the same trick to see that `_source` is a `java.util.LinkedHashMap`
+in the `_update` API:
+
+[source,js]
+---------------------------------------------------------
+POST /hockey/player/1/_update
+{
+  "script": "Debug.explain(ctx._source)"
+}
+---------------------------------------------------------
+// CONSOLE
+// TEST[continued catch:/painless_explain_error/]
+
+The response looks like:
+
+[source,js]
+---------------------------------------------------------
+{
+  "error" : {
+    "root_cause": ...,
+    "type": "illegal_argument_exception",
+    "reason": "failed to execute script",
+    "caused_by": {
+      "type": "script_exception",
+      "reason": "runtime error",
+      "caused_by": {
+        "type": "painless_explain_error",
+        "reason": "getClass=[java.util.LinkedHashMap] toString=[{gp=[26, 82, 1], last=gaudreau, assists=[17, 46, 0], first=johnny, goals=[9, 27, 1]}]" <1>
```

Comment:
Hmm - this is a LinkedHashMap but I'm not sure where it got the ordering from. I don't think this'll pass consistently as is because it probably relies on iteration order of some HashMap I don't have control over. I'll remove the assertion bit of this line and just keep it as an example.